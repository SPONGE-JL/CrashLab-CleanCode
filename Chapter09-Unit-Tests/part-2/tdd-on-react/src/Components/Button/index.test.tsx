import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import 'jest-styled-components';

import { Button } from './index';

describe('<Button />', () => {
  it('should render with style rules.', () => {
    const { container } = render(<Button label="Test Button" />);

    const label = screen.getByText('Test Button');
    expect(label).toBeInTheDocument();
    expect(label).toHaveStyleRule('color', '#FFFFFF');
    expect(label).toHaveStyleRule('font-size', '16px');

    const button = label.parentElement;
    expect(button).toHaveStyleRule('text-align', 'center');
    expect(button).toHaveStyleRule('cursor', 'pointer');
    expect(button).toHaveStyleRule('padding', '10px 20px');
    expect(button).toHaveStyleRule('border-radius', '8px');

    const defaultBackgroundColor = '#304FFE';
    const defaultHoverColor = '#1639FE';
    expect(button).toHaveStyleRule('background-color', defaultBackgroundColor);
    expect(button).toHaveStyleRule('background-color', defaultHoverColor, { modifier: ':hover' });

    expect(container).toMatchSnapshot();
  });

  it('renders changed "backgroundColor" and "hoberColor" with Props.', () => {
    const backgroundColor = '#FF1744';
    const hoverColor = '#F01440';
    render(<Button label="Button Test" backgroundColor={backgroundColor} hoverColor={hoverColor} />);

    const button = screen.getByText('Button Test').parentElement;
    expect(button).toHaveStyleRule('background-color', backgroundColor);
    expect(button).toHaveStyleRule('background-color', hoverColor, { modifier: ':hover' });
  });

  it('is a clickable button.', () => {
    const mockClick = jest.fn();
    render(<Button label="Button Test" onClick={mockClick} />);

    const buttonLabel = screen.getByText('Button Test');
    expect(mockClick).toHaveBeenCalledTimes(0);
    fireEvent.click(buttonLabel);
    expect(mockClick).toHaveBeenCalledTimes(1);
  });
});
