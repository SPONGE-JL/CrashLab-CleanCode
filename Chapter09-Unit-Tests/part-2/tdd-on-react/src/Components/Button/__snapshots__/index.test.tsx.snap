// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render with style rules. 1`] = `
.c0 {
  text-align: center;
  cursor: pointer;
  padding: 10px 20px;
  border-radius: 8px;
  background-color: #304FFE;
}

.c0:hover {
  background-color: #1639FE;
}

.c0:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c1 {
  color: #FFFFFF;
  font-size: 16px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      Test Button
    </div>
  </div>
</div>
`;
